feat: introduce GenerateRel for lateral view and unnest operations

EpsilonPrime opened the pull request:
Add GenerateRel, a new relational operator that applies table functions to produce zero or more output rows per input row. This enables SQL LATERAL VIEW, EXPLODE, and UNNEST operations for flattening nested data structures.

## Changes

- Adds TableFunction message based on original plans for such a message
- Adds GenerateRel message which is the first use of table functions

## Key Design Decisions

1. **Dedicated TableFunction message**: Provides type safety and clear semantics for multi-row producing functions, following the pattern of ScalarFunction, WindowFunction, and AggregateFunction.

2. **preserve_on_empty flag**: Controls NULL handling when generators produce zero rows. Named to reflect behavior (row preservation) rather than SQL join terminology ("outer"). Equivalent to Spark's LATERAL VIEW OUTER or standard SQL's LEFT JOIN LATERAL.

3. **No automatic ordinal column**: Unlike ExpandRel which appends an i32 ordinal, GenerateRel relies on table functions to include position explicitly (e.g., posexplode includes pos field in output) when needed.

4. **TableFunction not in Expression**: Table functions produce multiple rows, incompatible with Expression semantics (single value). Only used in relational operators like GenerateRel.

## Comparison to ExpandRel

- ExpandRel: Fixed cardinality (plan time), for grouping sets/cube
- GenerateRel: Variable cardinality (runtime), for lateral view/unnest

## Related Work

Will be used to unfork the changes made in Apache Gluten PR [#574](https://github.com/apache/incubator-gluten/pull/574).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/substrait-io/substrait/917)
<!-- Reviewable:end -->

EpsilonPrime committed:
feat: introduce GenerateRel for lateral view and unnest operations

Add GenerateRel, a new relational operator that applies table functions to
produce zero or more output rows per input row. This enables SQL LATERAL VIEW,
EXPLODE, and UNNEST operations for flattening nested data structures.

- Add TableFunction message (lines 1222-1262)
  - New expression type for functions that produce multiple rows
  - Fields: function_reference, arguments, options, output_type
  - Output type must be a Struct defining generated columns
  - Used exclusively in relational contexts (not in Expression.rex_type)

- Add GenerateRel message (lines 559-639)
  - Applies table function to each input row
  - Fields: common, input, generator, preserve_on_empty, advanced_extension
  - Comprehensive inline documentation with examples
  - Property maintenance: distribution and orderedness not maintained
  - Output schema: [input_fields..., generated_fields...]
  - Comparison to ExpandRel included in comments

- Add GenerateRel to Rel union (line 677)
  - Field number 23 in Rel.rel_type oneof

- site/docs/relations/logical_relations.md
  - Complete Generate Operation section with signature table
  - Common table functions reference (explode, posexplode, unnest)
  - Empty collection handling (preserve_on_empty flag)
  - Detailed examples: array explode, posexplode, map explode
  - Comparison table with ExpandRel

- site/docs/expressions/table_functions.md
  - Comprehensive rewrite of table functions documentation
  - TableFunction message field descriptions
  - Differences from scalar/window/aggregate functions
  - Usage examples with GenerateRel
  - Extension mechanism explanation

1. **Dedicated TableFunction message**: Provides type safety and clear
   semantics for multi-row producing functions, following the pattern of
   ScalarFunction, WindowFunction, and AggregateFunction.

2. **preserve_on_empty flag**: Controls NULL handling when generators produce
   zero rows. Named to reflect behavior (row preservation) rather than SQL
   join terminology ("outer"). Equivalent to Spark's LATERAL VIEW OUTER or
   standard SQL's LEFT JOIN LATERAL.

3. **No automatic ordinal column**: Unlike ExpandRel which appends an i32
   ordinal, GenerateRel relies on table functions to include position
   explicitly (e.g., posexplode includes pos field in output).

4. **TableFunction not in Expression**: Table functions produce multiple rows,
   incompatible with Expression semantics (single value). Only used in
   relational operators like GenerateRel.

- ExpandRel: Fixed cardinality (plan time), for grouping sets/cube
- GenerateRel: Variable cardinality (runtime), for lateral view/unnest

Based on Apache Gluten PR #574 but with significant improvements:
- Well-documented fields (Gluten had minimal documentation)
- Dedicated TableFunction type (vs generic Expression)
- preserve_on_empty flag for outer semantics
- Comprehensive examples and property maintenance rules

None. This is a purely additive change using a new field number in Rel.

- Protobuf compilation verified with protoc
- Follows all Substrait conventions (common field 1, input field 2,
  advanced_extension field 10)
- No trailing whitespace, proper line endings

benbellick reviewed (commented):
Very excited for this one, as it would be useful for modeling `Unnest` at my job. 

Couple things that I think are high priority: 
1. Could we introduce the schema changes for table functions in [simple_extension_schema.yaml](https://github.com/substrait-io/substrait/blob/main/text/simple_extensions_schema.yaml)? 
2. Could we introduce some proper examples into something like `functions_explode.yaml` to show these definitions? This also helps ensure that downstream libraries are properly handling these things. 
3. For my particular use case, I would love to be able to represent [queries like](https://www.db-fiddle.com/f/c12oDn2H6WknH1NtTXoVx8/0) `SELECT UNNEST(ARRAY[1,2]), UNNEST(ARRAY[4,5,6])` which has a zipping behavior rather than cartesian product. What would be the best way to represent these? (One way I can think of is as variadic unnest call which has zipping behavior.)

All in all this will be great and I'm happy to help implement them in the downstream libraries.

westonpace reviewed (commented):
I have a general question.  There are (at least) two ways in which table functions are utilized.  I'm not entirely sure if these are two different concepts or the same concept.

# Explode / lateral join / cross apply

I believe `GenerateRel` captures these cases.  The input to the table function is the output of another relation.  The arguments will (typically) include field references (or scalar functions applied to field references).

# From

The other way in which table functions are applied is by using them directly in the FROM clause.  For example `SELECT * FROM myfunc(7)`.  For example see [postgres](https://www.postgresql.org/docs/7.3/xfunc-tablefunctions.html).  In this case the arguments are (typically? always?) all literals.

Is the `TableFunction` message here intended to eventually be usable in both contexts?  This PR itself only adds support for the first (which is fine).  Still, it seems like message itself wouldn't have to change.  I think the only change I can imagine is adding a new source rel that has no input rels and specifies the table function invocation.

EpsilonPrime committed:
review notes

EpsilonPrime committed:
examples

EpsilonPrime committed:
non-variadic unnest

EpsilonPrime committed:
syntax fixes

EpsilonPrime committed:
fix coverage test

EpsilonPrime committed:
table function tests

EpsilonPrime committed:
better test case parser error messages

benbellick commented on proto/substrait/algebra.proto:
```diff
@@ -1690,6 +1726,54 @@ message Expression {
   }
 }
 
+// TableExpression represents expressions that produce zero or more rows,
```
If `TableExpression` only makes sense in the context of `GenerateRel`, should we move this proto definition to be within the scope of `GenerateRel`?

benbellick reviewed (commented):
I haven't done a complete review yet but just dumping my comments so far. Hopefully will get a chance to do rest later. Thanks!

EpsilonPrime committed:
added the description of table function test format

EpsilonPrime committed:
use any1 instead of convenient parameter names

yongchul commented on extensions/functions_table_generic.yaml:
```diff
@@ -0,0 +1,109 @@
+
+      For each input row, this function generates one output row for each element
+      in the input array. The output is a struct with two fields:
+      - 'pos': 0-indexed position of the element in the array (i64)
```
silly thing. some systems do use 1-indexed. do we want to have it as a function option?

EpsilonPrime replied:
Could conversion from 0-indexed to 1-indexed is something that could be handled by their Substrait layer?  If so, then we always canonicalize one way.

yongchul commented on proto/substrait/algebra.proto:
```diff
@@ -556,6 +556,41 @@ message ExpandRel {
+  //   This is equivalent to Spark's "LATERAL VIEW OUTER" or can be achieved in
+  //   standard SQL via LEFT JOIN LATERAL.
+  //
+  // Optional; defaults to false.
```
not: Optional`:`?

EpsilonPrime replied:
Leaving this as `Optional; defaults to false.` — it matches the existing convention used elsewhere in the file (e.g. the WindowFunction bound comments), so changing the punctuation here would make it less consistent.

yongchul commented on site/docs/expressions/table_functions.md:
```diff
@@ -1,8 +1,154 @@
+The `output_type` field must be a non-nullable `Struct` type where each field represents a column in the generated output rows. This explicitly defines the schema of rows produced by the table function.
+
+**Examples:**
+- `explode(array<i32>)` → `Struct{value: i32}`
+- `posexplode(array<string>)` → `Struct{pos: i64, value: string}`
+- `explode(map<string, i32>)` → `Struct{key: string, value: i32}`
```
wait.. do we have named struct? 😄

EpsilonPrime replied:
No named structs — Substrait structs are positional. The field names in those examples are illustrative only. I've added a note to the Output Type section spelling this out (356eb3c): generated columns map to the output schema by position, and names are just optional hints with no semantic meaning.

yongchul commented on site/docs/relations/logical_relations.md:
```diff
@@ -559,3 +559,105 @@ The operator that defines modifications of a database schema (CREATE/DROP/ALTER
+|----------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Inputs               | 1                                                                                                                                                              |
+| Outputs              | 1                                                                                                                                                              |
+| Property Maintenance | Distribution and orderedness are not maintained. The variable number of output rows per input breaks distribution properties, and row multiplication breaks ordering. |
```
input distribution is preserved as long as the input fields are intact and available -- which should be the case by definition.

EpsilonPrime replied:
You're right — the input fields are preserved unchanged on every output row, so distribution is maintained; only orderedness is broken by the row multiplication. Updated the Property Maintenance text and the ExpandRel comparison table accordingly (0cb8d3c).

yongchul commented on site/docs/relations/logical_relations.md:
```diff
@@ -559,3 +559,105 @@ The operator that defines modifications of a database schema (CREATE/DROP/ALTER
+| Inputs               | 1                                                                                                                                                              |
+| Outputs              | 1                                                                                                                                                              |
+| Property Maintenance | Distribution and orderedness are not maintained. The variable number of output rows per input breaks distribution properties, and row multiplication breaks ordering. |
+| Direct Output Order  | Input fields followed by generated fields. The output schema is: `[input_field_1, ..., input_field_N, generated_field_1, ..., generated_field_M]` where generated fields come from the table function's output type. Unlike ExpandRel which appends an i32 ordinal column, GenerateRel does NOT automatically add an ordinal column. Table functions that need position information (l ...
```
isn't output type is STRUCT? So we automatically flatten the struct in direct output order?

nielspardon commented:
@EpsilonPrime — #917 looks like it was closed by the stale-bot rather than on review grounds. I'd like to help carry `GenerateRel` forward as part of the Gluten↔Substrait consolidation. Would you prefer to reopen/continue it yourself, or are you happy for me to pick it up crediting your PR? (Minor: `generate = 23` now collides with `TopNRel`, so it'd become `= 25`.)

EpsilonPrime commented:
> @EpsilonPrime — #917 looks like it was closed by the stale-bot rather than on review grounds. I'd like to help carry `GenerateRel` forward as part of the Gluten↔Substrait consolidation. Would you prefer to reopen/continue it yourself, or are you happy for me to pick it up crediting your PR? (Minor: `generate = 23` now collides with `TopNRel`, so it'd become `= 25`.)

I can pick this back up tomorrow but if you have time to work on it today feel free to reverse the process.  Finishibg the Gluten cleanup would be nice.

nielspardon commented:
Alright, let me reopen this one. I'm going through the list of Gluten modifications currently and I'm in touch internally at IBM with the Gluten team on this. Let me leave this one to you and I'll carry on with the other items.

EpsilonPrime committed:
fix: adapt GenerateRel to current main after rebase

Resolve semantic conflicts from rebasing the GenerateRel branch onto main:

- proto: move GenerateRel to field number 25 in Rel.rel_type (23 and 24
  were taken by TopNRel and LateralJoinRel since the branch was authored).
- grammar: drop the duplicate structType rule/alternative; main already
  added structType to FuncTestCaseParser.g4. Regenerate the ANTLR parsers.
- coverage: register table functions under their extension URN (main
  migrated includes from file paths to URNs) and update the table_generic
  test cases to the URN include format.
- visitor: remove the duplicate visitDataType now provided by main.
- baseline: refresh tests/baseline.json for the merged registry counts.
- validator: add table_function textproto examples to the example
  validator so GenerateRel/TableFunction examples stay valid.

EpsilonPrime committed:
docs: address table-function review comments

Resolve the outstanding actionable review feedback on the GenerateRel PR:

- extensions: merge the two `explode` definitions into a single function
  with two impls (list and map), per the "function names should be unique"
  discussion (#931). This also removes the explode_dup rename from the
  coverage registry. Fold the "some systems call this unnest" note into the
  combined description so it applies to both forms.
- docs: correct the GenerateRel property maintenance -- distribution IS
  maintained because the input fields are preserved unchanged on every
  output row; only orderedness is broken by row multiplication.
- docs: mark TableFunction `arguments` as Optional (a table function may
  take no arguments).
- docs/proto: "input record" -> "input row" wording nits.
- tests: refresh baseline.json for the merged explode (num_table_functions
  4 -> 3).

EpsilonPrime commented on extensions/functions_table_generic.yaml:
```diff
@@ -0,0 +1,109 @@
+---
+urn: extension:io.substrait:functions_table_generic
+table_functions:
+  - name: explode
```
Good call — merged the array and map forms into a single `explode` function with two `impls` (0cb8d3c). That also drops the `explode_dup1` rename from the coverage registry.

EpsilonPrime commented on extensions/functions_table_generic.yaml:
```diff
@@ -0,0 +1,109 @@
+      Empty maps produce zero output rows. For outer semantics (producing a row
+      with NULLs when the map is empty), use GenerateRel's preserve_on_empty option.
+
+      Some systems call this unnest.
```
Done — with the two explode forms now merged into one function, the "some systems call this unnest" note lives in the shared description and applies to both (0cb8d3c).

EpsilonPrime commented on site/docs/expressions/table_functions.md:
```diff
@@ -1,8 +1,154 @@
+| Field | Type | Description | Required |
+|-------|------|-------------|----------|
+| function_reference | uint32 | Points to a function_anchor in the plan that references a table function in the extension YAML. 0 is a valid anchor/reference. | Required |
+| arguments | repeated FunctionArgument | Arguments bound to the table function (value, type, or enum arguments) | Required |
```
Agreed — a table function can take no arguments. Marked `arguments` as Optional (it's a `repeated` field, so it may be empty) (0cb8d3c).

EpsilonPrime commented on site/docs/expressions/table_functions.md:
```diff
@@ -1,8 +1,154 @@
+- **Scalar functions:** Produce exactly one value per input row
+- **Window functions:** Produce one value per input row (with access to window of rows)
+- **Aggregate functions:** Produce one value per group of input rows
+- **Table functions:** Produce zero or more rows per input record
```
Fixed — "input record" → "input row", and I applied the same wording fix to the matching GenerateRel proto comment (0cb8d3c).

EpsilonPrime committed:
docs: clarify that table function output structs are positional

Address review discussion on named structs: the field names shown in the
output_type examples (value, pos, key) are illustrative only. Substrait
structs are positional, so generated columns map to the output schema by
position; names are optional hints with no semantic meaning.

EpsilonPrime commented on site/docs/expressions/table_functions.md:
```diff
@@ -1,8 +1,154 @@
+| **Input scope** | Single row | Window of rows | Group of rows | Single row |
+| **Output cardinality** | Exactly 1 value | Exactly 1 value | Exactly 1 value | 0 or more rows |
+| **Output type** | Scalar type | Scalar type | Scalar type | Struct type |
+| **Can appear in Expression** | Yes | Yes (WindowFunction) | No | No |
```
Right — same point: any "names" here (or in make_struct-style functions) are positional hints, not part of the type. Clarified in the Output Type note added in 356eb3c.

nielspardon reviewed (commented):
Overall this looks like the right shape — an additive relation plus table functions declared through the normal extension mechanism rather than built-in generators. Four things to fix before merge, and three design questions worth settling while field numbers are still cheap to change.

## Blocking

**Nullability is self-contradictory.** No impl sets `nullability:`, so all seven default to MIRROR — and MIRROR makes `explode(list?<i32>)` derive `struct?<i32>`, which `table_functions.md` forbids ("must be a non-nullable `Struct` type"). A nullable collection column therefore has no legal `output_type`, and that's the common case.

**`preserve_on_empty` can't produce the NULLs it promises.** `posexplode` returns `struct<i64, any1>`, and this PR's own `posexplode.textproto` marks `pos` as `NULLABILITY_REQUIRED`. One rule fixes it: when `preserve_on_empty` is true, every field of `output_type` is nullable.

**The new kind isn't wired into existing machinery.** `tests/baseline.py` got its `num_table_functions` counter; six siblings didn't:

| Where | Missing |
|---|---|
| `text/dialect_schema.yaml` | `GENERATE`, `generate_relation`, `supported_table_functions` |
| `text/simple_extensions_schema.yaml` | `deprecated`, `metadata` — so a table function can never be deprecated *(details inline)* |
| `tests/coverage/coverage.py` | `_result_is_nullable` not extended for `MultiRowResult` — no nullable-input table test can pass |
| `tests/type/test_type_grammar.py` | `table_functions` in `iter_type_expressions` — new type strings are never grammar-checked |
| `site/docs/extensions/index.md` | two "scalar, aggregate, and window" enumerations (lines 164, 175) |
| `site/examples/proto-textformat/README.md` | `table_function/` missing from the Directories index |

I'd rather these land with the feature. The dialect schema has drifted before (`TOP_N`, `EXTENSION_*`), but that argues for not extending the drift — and the function kind has no precedent at all: nothing anywhere can express table-function support until this lands.

<details>
<summary>Concrete edits for the five that aren't in the diff</summary>

**`text/dialect_schema.yaml`**

```yaml
# supported_relations enum + the parallel oneOf list
- GENERATE
- $ref: "#/definitions/generate_relation"

# definitions, alongside expand_relation
  generate_relation:
    type: object
    additionalProperties: false
    properties:
      relation: {const: GENERATE}
      metadata: {type: object}
    required: [relation]

# alongside supported_scalar_functions
  supported_table_functions:
    type: array
    items: {$ref: "#/definitions/dialect_function"}
```

`expand_relation` carries `field_types` as its per-relation capability; the analogue here would be whether the dialect supports `preserve_on_empty`, which engines genuinely differ on. One open question, not a request: does `supported_expressions` need a table-expression member? Probably not if `supported_table_functions` exists — but if `TableExpression` keeps its wrapper shape, a dialect can't say it supports the container but not a given function.

**`tests/coverage/coverage.py`** — `TestCase.result` gained `MultiRowResult`, but `_result_is_nullable` still only handles `CaseLiteral`, so it returns False for every table result regardless of what the file declares. That makes any nullable input unsatisfiable — both spellings fail identically:

```
explode([1, 2]::list?<i32>) = [(1), (2)]::struct?<i32>
→ MIRROR violation: output should be nullable because an input is nullable
```

The grammar accepts `structType`'s `isnull` and `visitMultiRowResult` keeps the `?` — it's just dropped here, since `MultiRowResult` has no `nullable`/`is_nullable()`. DECLARED_OUTPUT and DISCRETE have the same gap. Once the YAML sets nullability explicitly, this is what stands between us and a test for the nullable case.

**`tests/type/test_type_grammar.py`** — one-line fix: `iter_type_expressions` walks `scalar_functions`/`aggregate_functions`/`window_functions` but not `table_functions`. The JSON schema can't cover it either (`$defs/type` is `oneOf: [string, object]` with no pattern, so any string validates). Latent rather than live, since the current strings do parse — but `return: struct<value: T>`, the form the new docs recommend, would have been caught immediately.

**`site/docs/extensions/index.md`** — line 14 has listed "Table Functions" as a simple-extension kind for a while, but 164 still says "(scalar, aggregate, and window functions)" and 175 still scopes `deprecated` to "(scalar, aggregate, and window)". This is the PR that makes the kind real; 175 can only become true once the schema def gains `deprecated`.

**`site/examples/proto-textformat/README.md`** — the new `table_function/` directory isn't in the Directories index, which is also the only place a reader learns the one-message-class-per-directory convention.
</details>

**The test surface should arrive complete.** `explode(map)` ships with no test, and the new format can't express either an expected error or a nested value in a result row — both of which these functions specifically need. Details inline.

## Design questions

**Relationship to `LateralJoinRel`?** It's already on `main`, evaluates its right input once per left row, and its `JoinType` INNER/LEFT is exactly `preserve_on_empty` — but the new docs compare only against `ExpandRel`, twenty lines below `LateralJoinRel`'s own section. A producer facing `LATERAL VIEW OUTER EXPLODE(tags)` now has two encodings and no rule for choosing. A paragraph saying which to emit would settle it. Two knock-ons: a bool can't grow the way the enum can (LEFT_SEMI/ANTI/SINGLE/MARK all have Generate analogues), and `LateralJoinRel` defines a `rel_anchor`/`OuterReference` scope where `GenerateRel` says nothing about the generator's reference scope.

**`TableExpression` shape.** Single-member oneof, one reference in the tree, and the non-composability it exists to express is prose rather than enforced. `AggregateFunction` is the precedent for the same situation — a call that can't appear inside `Expression`, modelled as a bare top-level message. As written, every SDK adds a whole expression-kind axis to reach one leaf, and `TableExpression{}` with no generator is schema-valid. Hoisting `TableFunction`, or putting the oneof on `GenerateRel` directly (house style: `ReadRel.read_type`, `WriteRel`, `DdlRel`), avoids both.

**Variadic return arity.** `returnValue` can't express a struct whose arity tracks the argument count, so `unnest` enumerates impls and stops at four. Real generators exceed that — Spark `stack(n, v1..vk)` yields `ceil(k/n)` columns unbounded. Fine to defer the general fix, but the docs should say the arity list is an enumeration, not a model limit.

<details>
<summary><b>Gluten migration check</b> — nothing inexpressible; items 1 and 2 are what block it</summary>

Gluten ships a vendored `GenerateRel`, so I checked what a migration would need:

| Gluten | This PR |
|---|---|
| `repeated Expression child_output` | `RelCommon.emit` — it's only ever field selections (producer emits `makeSelection(childIndex)`; the Velox consumer asserts field-only), so `emit` covers it and the field shouldn't be carried over |
| `bool outer` | `preserve_on_empty` |
| `GenerateParameters:` string hints in `advanced_extension` | derivable from the function anchor — also retires a positional convention where the consumer assumes the generator input is the child projection's last column |

So the migration is a simplification. The two blockers: Spark array columns are nullable (item 1), and `LATERAL VIEW OUTER POSEXPLODE` needs a NULL position (item 2).
</details>

Worth a short version of that table in the PR body — AGENTS.md asks for the per-library note on `proto/` changes.

nielspardon commented on extensions/functions_table_generic.yaml:
```diff
@@ -0,0 +1,102 @@
+          - name: input
+            value: list<any1>
+            description: The array to explode.
+        return: struct<any1>
```
No impl here sets `nullability:`, so all seven default to MIRROR — which makes `explode(list?<i32>)` derive `struct?<i32>`, a nullable struct that `table_functions.md` forbids. So a nullable collection column has no legal `output_type`.

```yaml
        return: struct<any1>
        nullability: DECLARED_OUTPUT
```

Separately: the descriptions say what an *empty* collection does, never a NULL one. Stating that NULL behaves as empty would match what engines do and let `preserve_on_empty` cover it.

nielspardon commented on extensions/functions_table_generic.yaml:
```diff
@@ -0,0 +1,102 @@
+      When multiple lists are provided, produces rows by expanding the lists in parallel,
+      similar to a zip operation. For example, unnest([1,2], [3,4]) produces rows: (1,3), (2,4).
+
+      If lists have different lengths, shorter lists are padded with NULL values to match
```
"shorter lists are padded with NULL values" can't be satisfied by `struct<any1, any2>` under MIRROR over non-nullable lists — the padded position can't hold NULL. Same fix as above.

Two notes: no test exercises unequal lengths (every case zips equal-length lists), and it's worth saying the four arities are an enumeration rather than a model limit, since `returnValue` can't yet express argument-dependent arity.

nielspardon commented on proto/substrait/algebra.proto:
```diff
@@ -627,6 +627,41 @@ enum FetchMode {
+  //   standard SQL via LEFT JOIN LATERAL.
+  //
+  // Optional; defaults to false.
+  bool preserve_on_empty = 4;
```
This sets the generated fields to NULL, but nothing promotes them to nullable — and `posexplode.textproto` in this PR declares `pos` as `NULLABILITY_REQUIRED`, so `preserve_on_empty: true` writes NULL into a field whose type forbids it. Under MIRROR over non-nullable lists that's the derived default, not an edge case.

Could we state the rule here and in Empty Collection Handling — when `preserve_on_empty` is true, every field of `output_type` is nullable? Without it, two consumers derive different schemas for the same plan.

nielspardon commented on text/simple_extensions_schema.yaml:
```diff
@@ -376,3 +380,37 @@ $defs:
               type: string
               enum: [STREAMING, PARTITION]
+
+  tableFunction:
```
This def omits `deprecated` (function *and* impl level), `metadata`, and impl-level `description`, all of which the other three function defs carry. With `additionalProperties: false` those are hard failures, so a table function can never be deprecated — awkward with the `functions_aggregate_decimal_output` deprecation in flight. The missing impl `description` also makes `generate_function_docs.py`'s per-impl rendering unreachable here.

Since all four defs share an identical impl block, a shared `#/$defs/functionImpl` would put the next added property in one place instead of four.

nielspardon commented on grammar/FuncTestCaseParser.g4:
```diff
@@ -78,6 +103,18 @@ aggFuncTestCase
+    : functionName=identifier OParen arguments CParen ( OBracket funcOptions CBracket )? Eq multiRowResult
+    ;
+
+multiRowResult
```
Two gaps versus the other two formats, both cheaper to close now than once files depend on the shape:

- **No expected errors.** `multiRowResult` has no `substraitError` alternative, so `is_return_type_error()` is permanently False here. That's exactly where these functions are least specified — `explode` on NULL, `posexplode` position overflow, `unnest` on unequal lengths.
- **No nested result values.** `rowTuple` uses `literal` rather than `compoundLiteral`, so `explode([[1,2],[3,4]]::list<list<i32>>) = [([1, 2]), ([3, 4])]::struct<list<i32>>` doesn't parse — the nested-flattening case the feature exists for. `literalStruct`/`literalMap` already use `compoundLiteral`.

`rowTuple` also duplicates `columnValues`; `multiRowResult : literalList DoubleColon structType ;` reuses both and lets `MultiRowResult` collapse into `CaseLiteral`.

nielspardon commented on tests/baseline.json:
```diff
@@ -1,16 +1,17 @@
-    "num_covered_function_variants": 245
+    "total_test_count": 1257,
+    "num_function_variants": 509,
+    "num_covered_function_variants": 255
```
`explode(map<any1, any2>)` is the one uncovered variant of the seven. It slips through because the accepted gap is rewritten here (253 → 254) and `validate_against` compares with `>`, so an equal gap passes — the exact case its own message targets ("New functions should be added along with test cases that illustrate their behavior").

It's also the overload whose iteration order the YAML declares unspecified, so the most worth pinning down. I know the baseline on `main` was already stale — which is why the deltas look larger than they are — but I'd rather a legacy gap didn't absorb a new one.

nielspardon commented on tests/coverage/extensions.py:
```diff
@@ -81,6 +81,9 @@ def get_short_type(long_type):
 
         if short_type is None:
+            # For struct types, preserve the full parameterized type
+            if long_type.startswith("struct<"):
```
I think both hunks can come out. `struct` is already in `type_to_short_type` (`test_build_type_to_short_type` pins it), so the existing param-stripping path normalizes `struct<i64, any1>` → `struct` and the original `is_same_type` matched it:

| Reverted | Result |
|---|---|
| `struct<` early return only | green |
| `is_same_type` widening only | 18 failures |
| **both** | green, byte-identical `baseline.json` |

So the pair adds no capability — but the widening isn't table-scoped, it drops parameters for all 509 overloads. `is_same_type('dec<38,0>', 'dec<10,2>')` and `('list<i32>', 'list<str>')` were False before and are True now; unreachable today only because struct is the sole type that keeps its parameters.

Live cost: `posexplode(...) = [(1, 2, 3), ('a', 'b', 'c')]::struct<i64,i32>` — wrong arity *and* element types — is accepted and counted as coverage of the `struct<i64, any1>` variant.

nielspardon commented on tests/coverage/visitor.py:
```diff
@@ -546,6 +614,22 @@ def visitSubstraitError(self, ctx: FuncTestCaseParser.SubstraitErrorContext):
             return SubstraitError("undefined")
         return SubstraitError("unknown_error")
 
+    def visitMultiRowResult(self, ctx: FuncTestCaseParser.MultiRowResultContext):
```
`rows` and `type` are stored independently and never cross-checked, so `[(1, 2, 3)]::struct<i32>`, `[()]::struct<i32>`, and rows of differing widths in one result all validate clean. `visitRowTuple` also drops each literal's type (`value, _ = self.visitLiteral(literal)`), so the rows are untyped strings. Combined with the `is_same_type` change, the format currently asserts only function name, argument base types and argument count.

Also unguarded: `[(P1Y2M3DT4H5M6S)]::struct<icompound>` parses, then raises `TypeError: cannot unpack non-iterable NoneType object` — `visitLiteral` has no `IntervalCompoundLiteral` branch. Since `load_all_testcases` walks the whole tree, one such file aborts the run with no file or line attribution.

nielspardon commented on site/docs/expressions/table_functions.md:
```diff
@@ -1,8 +1,156 @@
+      - args:
+          - name: input
+            value: list<T>
+        return: struct<value: T>
```
Five things here will send implementers the wrong way:

- `return: struct<value: T>` — named fields are `nstruct<>`, parameters are `any1` not `T`; this fails the repo's own type grammar, and contradicts line 34 ("structs are positional, not named ... illustrative only").
- The snippet is inline, so neither `check-jsonschema` nor the type-grammar test sees it — an `--8<--` include of the real YAML can't drift.
- "Must be marked as table-valued in function metadata" — no such marker; the `table_functions:` key is the marker, and `metadata` is in fact rejected by the new def.
- `array<T>` here and in the relation docs — the type class is `list<T>` (`array<i32>` doesn't parse).
- The `unnest` row calls it a synonym for `explode` over one array; the YAML defines parallel zip over up to four lists with NULL padding. Implementing from the docs would mis-execute the multi-argument cases this PR ships tests for.

nielspardon commented on tests/README.md:
````diff
@@ -339,3 +372,22 @@ In this example:
+
+```code
+### SUBSTRAIT_TABLE_TEST: v1.0
+### SUBSTRAIT_INCLUDE: '/extensions/functions_table_generic.yaml'
````
Looks like a rebase leftover — this example uses the pre-URN include form and doesn't parse: `mismatched input ''/extensions/functions_table_generic.yaml'' expecting ExtensionUrn`. Every other include here uses the URN form, and the three real `.test` files get it right. `### SUBSTRAIT_TABLE_TEST: V1` above it is also invalid — `FormatVersion` is lowercase.

nielspardon commented on tests/test_proto_example_validator.py:
```diff
@@ -64,3 +64,18 @@ def test_validate_field_references():
+        "simple_explode.textproto": algebra_pb2.TableExpression.TableFunction,
+    }
+    for textproto_file in examples_dir.glob("*.textproto"):
+        message_class = message_by_file[textproto_file.name]
```
The filename→message dict raises `KeyError` for any new `.textproto` in `table_function/`, and silently does less work if one is renamed. Root cause is that this directory mixes `GenerateRel` and `TableExpression.TableFunction` where the siblings hold one class each — splitting into `generate_rel/` and `table_function/` restores that and lets all four tests collapse into one parametrized case.

nielspardon commented on grammar/FuncTestCaseParser.g4:
```diff
@@ -6,15 +6,33 @@ options {
 }
 
 doc
```
Curiosity, not a request: was splitting `doc`/`header`/`testGroup` per-kind deliberate? Extending the existing token alternation would have been a much smaller diff — but the split also makes the corpus parse noticeably faster, presumably because the per-kind `doc` resolves on two tokens instead of ALL(*) lookahead at every group boundary. If that's why, worth a comment in the grammar, since the obvious future refactor would undo it.
